### PR TITLE
log(error): logging the error messages

### PR DESCRIPTION
**Pull request by ispeakc0de**
Signed-off-by: shubhamchaudhary <shubham@chaosnative.com>

**What this PR does / why we need it**:
- logging the error msg when unable to get chaosengine
- If chaosengine was not there then we are trying to access it so it panic mode as it tries to access attributes of nil object. Added the check for the same.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: bin/runner.go [hunk 1]
```diff
@@ -22,11 +22,16 @@ func main() {
 	clients := utils.ClientSets{}
 	// Getting kubeConfig and Generate ClientSets
 	if err := clients.GenerateClientSetFromKubeConfig(); err != nil {
-		log.Fatalf("Unable to create ClientSets, error: %v", err)
+		log.Errorf("Unable to create ClientSets, error: %v", err)
```
Comment (rahulchheda) [inline bin/runner.go:25]: Please don't capitalise initial letters of log message

File: pkg/utils/environment.go [hunk 2]
```diff
@@ -27,13 +27,13 @@ func (engineDetails *EngineDetails) GetOsEnv() *EngineDetails {
 }
 
 // GetEngineUID get the chaosengine UID
-func (engineDetails *EngineDetails) GetEngineUID(clients ClientSets) *EngineDetails {
+func (engineDetails *EngineDetails) GetEngineUID(clients ClientSets) error {
```
Comment (rahulchheda) [inline pkg/utils/environment.go:30]: What is the purpose of this function if it just returns error

**Review by ksatchit — DISMISSED** [review]

File: bin/runner.go (hunk 1 above)
Comment (ispeakc0de) [reply bin/runner.go]: updated

**Review by ksatchit — APPROVED** [review]

File: bin/runner.go [hunk 3]
```diff
@@ -22,11 +22,16 @@ func main() {
...
 	if err := clients.GenerateClientSetFromKubeConfig(); err != nil {
-		log.Fatalf("Unable to create ClientSets, error: %v", err)
+		log.Errorf("unable to create ClientSets, error: %v", err)
+		return
 	}
 	// Fetching all the ENVs passed from the chaos-operator
 	// create and initialize the experimentList
-	experimentList := engineDetails.GetOsEnv().GetEngineUID(clients).CreateExperimentList()
+	if err := engineDetails.SetEnv().SetEngineUID(clients); err != nil {
+		log.Errorf("unable to get ChaosEngineUID, err: %v", err)
```
Comment (rahulchheda) [inline bin/runner.go:31]:
```suggestion
		log.Errorf("unable to get ChaosEngineUID, error: %v", err)
```

File: bin/runner.go [hunk 4]
```diff
@@ -50,27 +56,27 @@ func main() {
 		}
 		// check the existance of chaosexperiment inside the cluster
 		if err := experiment.HandleChaosExperimentExistence(engineDetails, clients); err != nil {
-			log.Errorf("Unable to get ChaosExperiment Name: %v, in namespace: %v, error: %v", experiment.Name, experiment.Namespace, err)
+			log.Errorf("unable to get ChaosExperiment Name: %v, in namespace: %v, error: %v", experiment.Name, experiment.Namespace, err)
```
Comment (rahulchheda) [inline bin/runner.go:59]:
```suggestion
			log.Errorf("unable to get ChaosExperiment name: %v, in namespace: %v, error: %v", experiment.Name, experiment.Namespace, err)
```

File: pkg/utils/environment.go [hunk 5]
```diff
@@ -10,8 +10,8 @@ import (
 	"github.com/litmuschaos/chaos-runner/pkg/log"
 )
 
-// GetOsEnv adds the ENV's to EngineDetails
-func (engineDetails *EngineDetails) GetOsEnv() *EngineDetails {
+// SetEnv adds the ENV's to EngineDetails
+func (engineDetails *EngineDetails) SetEnv() *EngineDetails {
```
Comment (rahulchheda) [inline pkg/utils/environment.go:14]:
```suggestion
func (engineDetails *EngineDetails) SetEngineDetails() *EngineDetails {
```

fn name seems to be little misleading.

File: tests/runner_test.go [hunk 6]
```diff
@@ -218,26 +211,39 @@ var _ = Describe("BDD on chaos-runner", func() {
...
-
-			//Fetching engine-nginx-runner pod
-			runner, err := k8sClientSet.CoreV1().Pods("litmus").Get("engine-nginx-runner", metav1.GetOptions{})
-			Expect(err).To(BeNil())
-			Expect(string(runner.Status.Phase)).To(Or(Equal("Running"), Equal("Succeeded")))
+			log.Info("chaos engine created")
+
+			err = retry.
+				Times(uint(180 / 2)).
+				Wait(time.Duration(2) * time.Second).
```
Comment (rahulchheda) [inline tests/runner_test.go:223]:
```suggestion
				Wait(time.Duration(2 * time.Second)).
```

Comment (rahulchheda) [reply tests/runner_test.go:223]: I hope everything is tested here.

File: tests/runner_test.go [hunk 7]
```diff
@@ -218,26 +211,39 @@ var _ = Describe("BDD on chaos-runner", func() {
...
+				Try(func(attempt uint) error {
+					pod, err := clients.KubeClient.CoreV1().Pods("litmus").Get("engine-nginx-runner", metav1.GetOptions{})
+					if err != nil {
+						return errors.Errorf("unable to get chaos-runner pod, err: %v", err)
+					}
+					if pod.Status.Phase != "Running" && pod.Status.Phase != "Succeeded" {
```
Comment (rahulchheda) [inline tests/runner_test.go:229]: Please create typed consts for different phases.

File: tests/runner_test.go [hunk 8]
```diff
@@ -25,105 +25,97 @@ import (
...
+	Expect(err).To(BeNil(), "the chaos-operator is not in running state")
+	log.Info("Chaos-Operator is in running state")
 
-	err = exec.Command("kubectl", "apply", "-f", "https://hub.litmuschaos.io/api/chaos/master?file=charts/generic/experiments.yaml", "-n", "litmus").Run()
-	if err != nil {
-		klog.Infof("Unable to create Pod-Delete Experiment, due to error: %v", err)
-	}
+	err = exec.Command("kubectl", "apply", "-f", "https://hub.litmuschaos.io/api/chaos/master?file=charts/generic/pod-delete/experiment.yaml", "-n", "litmus").Run()
+	Expect(err).To(BeNil(), "unable to create Pod-Delete Experiment")
+	log.Info("pod-delete CR created")
```
Comment (rahulchheda) [inline tests/runner_test.go:114]:
```suggestion
	log.Info("pod-delete ChaosExperiment created")
```

File: tests/runner_test.go [hunk 9]
```diff
@@ -25,105 +25,97 @@ import (
...
-	operator, _ := k8sClientSet.CoreV1().Pods("litmus").List(metav1.ListOptions{LabelSelector: "name=chaos-operator"})
-	for _, v := range operator.Items {
+	err = exec.Command("kubectl", "apply", "-f", "https://raw.githubusercontent.com/litmuschaos/chaos-operator/master/deploy/operator.yaml").Run()
+	Expect(err).To(BeNil(), "unable to create Chaos-operator")
+
+	log.Info("Chaos-Operator created")
+
+	err = retry.
+		Times(uint(180 / 2)).
+		Wait(time.Duration(2) * time.Second).
```
Comment (rahulchheda) [inline tests/runner_test.go:95]:
```suggestion
		Wait(time.Duration(2 * time.Second)).
```

File: tests/runner_test.go [hunk 10]
```diff
@@ -25,105 +25,97 @@ import (
...
-		break
-	}
+	Expect(err).To(BeNil(), "the chaos-operator is not in running state")
+	log.Info("Chaos-Operator is in running state")
 
-	err = exec.Command("kubectl", "apply", "-f", "https://hub.litmuschaos.io/api/chaos/master?file=charts/generic/experiments.yaml", "-n", "litmus").Run()
-	if err != nil {
-		klog.Infof("Unable to create Pod-Delete Experiment, due to error: %v", err)
-	}
+	err = exec.Command("kubectl", "apply", "-f", "https://hub.litmuschaos.io/api/chaos/master?file=charts/generic/pod-delete/experiment.yaml", "-n", "litmus").Run()
```
Comment (rahulchheda) [inline tests/runner_test.go:112]:
```suggestion
        By("Installing Pod Delete Experiment")
	err = exec.Command("kubectl", "apply", "-f", "https://hub.litmuschaos.io/api/chaos/master?file=charts/generic/pod-delete/experiment.yaml", "-n", "litmus").Run()
```

File: tests/runner_test.go [hunk 11]
```diff
@@ -25,105 +25,97 @@ import (
...
-
-	if err != nil {
-		Expect(err).To(BeNil(), "failed to get k8sClientSet")
-	}
+	k8sClientSet, err := k8s.GenerateK8sClientSet(config)
+	Expect(err).To(BeNil(), "failed to get k8sClientSet")
 
-	litmusClientSet, err = chaosClient.NewForConfig(config)
+	litmusClientSet, err := litmus.GenerateLitmusClientSet(config)
+	Expect(err).To(BeNil(), "failed to get litmusClientSet")
```
Comment (rahulchheda) [inline tests/runner_test.go:68]:
```suggestion
	Expect(err).To(BeNil(), "failed to generate litmusClientSet")
```

File: tests/runner_test.go [hunk 12]
```diff
@@ -25,105 +25,97 @@ import (
...
-		Expect(err).To(BeNil(), "failed to get config")
-	}
-
-	k8sClientSet, err = kubernetes.NewForConfig(config)
-
-	if err != nil {
-		Expect(err).To(BeNil(), "failed to get k8sClientSet")
-	}
+	k8sClientSet, err := k8s.GenerateK8sClientSet(config)
+	Expect(err).To(BeNil(), "failed to get k8sClientSet")
```
Comment (rahulchheda) [inline tests/runner_test.go:65]:
```suggestion
	Expect(err).To(BeNil(), "failed to generate k8sClientSet")
```

File: tests/runner_test.go [hunk 13]
```diff
@@ -25,105 +25,97 @@ import (
...
 	RegisterFailHandler(Fail)
 	RunSpecs(t, "BDD test")
 }
+func init() {
+	flag.StringVar(&kubeconfig, "kubeconfig", os.Getenv("HOME")+"/.kube/config", "path to kubeconfig to invoke kubernetes API calls")
+}
 
 var _ = BeforeSuite(func() {
 
+	// Getting kubeConfig and Generate ClientSets
```
Comment (rahulchheda) [inline tests/runner_test.go:59]:
```suggestion
	// Getting kubeconfig and generate clientSets
```

File: pkg/utils/watchJob.go [hunk 14]
```diff
@@ -107,7 +107,7 @@ func (engineDetails EngineDetails) DeleteJobAccordingToJobCleanUpPolicy(experime
 			PropagationPolicy: &deletePolicy,
 		})
 		if deleteJob != nil {
-			return "", errors.Errorf("Unable to delete ChaosExperiment Job Name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
+			return "", errors.Errorf("unable to delete ChaosExperiment Job Name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:110]:
```suggestion
			return "", errors.Errorf("unable to delete ChaosExperiment Job name: %v, in namespace: %v, error: %v", experiment.JobName, experiment.Namespace, err)
```

File: pkg/utils/watchJob.go [hunk 15]
```diff
@@ -64,7 +64,7 @@ func (expDetails *ExperimentDetails) GetChaosResult(engineDetails EngineDetails,
 	resultName := GetResultName(engineDetails.Name, expDetails.Name, expDetails.InstanceID)
 	expResult, err := clients.LitmusClient.LitmuschaosV1alpha1().ChaosResults(engineDetails.EngineNamespace).Get(resultName, metav1.GetOptions{})
 	if err != nil {
-		return nil, errors.Errorf("Unable to get ChaosResult Name: %v in namespace: %v, error: %v", resultName, engineDetails.EngineNamespace, err)
+		return nil, errors.Errorf("unable to get ChaosResult Name: %v in namespace: %v, error: %v", resultName, engineDetails.EngineNamespace, err)
```
Comment (rahulchheda) [inline pkg/utils/watchJob.go:67]:
```suggestion
		return nil, errors.Errorf("unable to get ChaosResult name: %v in namespace: %v, error: %v", resultName, engineDetails.EngineNamespace, err)
```

File: pkg/utils/watchChaosContainer.go [hunk 16]
```diff
@@ -21,7 +21,7 @@ func GetChaosPod(expDetails *ExperimentDetails, clients ClientSets) (*corev1.Pod
 		Try(func(attempt uint) error {
 			chaosPodList, err = clients.KubeClient.CoreV1().Pods(expDetails.Namespace).List(metav1.ListOptions{LabelSelector: "job-name=" + expDetails.JobName})
 			if err != nil || len(chaosPodList.Items) == 0 {
-				return errors.Errorf("Unable to get the chaos pod, error: %v", err)
+				return errors.Errorf("unable to get the chaos pod, error: %v", err)
```
Comment (rahulchheda) [inline pkg/utils/watchChaosContainer.go:24]:
```suggestion
				return fmt.Errorf("unable to get the chaos pod, error: %w", err)
```

File: pkg/utils/watchChaosContainer.go [hunk 17]
```diff
@@ -46,7 +46,7 @@ func GetChaosContainerStatus(experimentDetails *ExperimentDetails, clients Clien
 
 	pod, err := GetChaosPod(experimentDetails, clients)
 	if err != nil {
-		return false, errors.Errorf("Unable to get the chaos pod, error: %v", err)
+		return false, errors.Errorf("unable to get the chaos pod, error: %v", err)
```
Comment (rahulchheda) [inline pkg/utils/watchChaosContainer.go:49]:
```suggestion
		return false, fmt.Errorf("unable to get the chaos pod, error: %w", err)
```

File: pkg/utils/watchChaosContainer.go [hunk 18]
```diff
@@ -70,7 +70,7 @@ func GetChaosContainerStatus(experimentDetails *ExperimentDetails, clients Clien
 			Try(func(attempt uint) error {
 				pod, err := GetChaosPod(experimentDetails, clients)
 				if err != nil {
-					return errors.Errorf("Unable to get the chaos pod, error: %v", err)
+					return errors.Errorf("unable to get the chaos pod, error: %v", err)
```
Comment (rahulchheda) [inline pkg/utils/watchChaosContainer.go:73]:
```suggestion
					return fmt.Errorf("unable to get the chaos pod, error: %w", err)
```

File: pkg/utils/watchChaosContainer.go [hunk 19]
```diff
@@ -103,12 +103,12 @@ func (engineDetails EngineDetails) WatchChaosContainerForCompletion(experiment *
 		var expStatus ExperimentStatus
 		chaosPod, err := GetChaosPod(experiment, clients)
 		if err != nil {
-			return errors.Errorf("Unable to get the chaos pod, error: %v", err)
+			return errors.Errorf("unable to get the chaos pod, error: %v", err)
```
Comment (rahulchheda) [inline pkg/utils/watchChaosContainer.go:106]:
```suggestion
			return fmt.Errorf("unable to get the chaos pod, error: %w", err)
```

File: pkg/utils/watchChaosContainer.go [hunk 20]
```diff
@@ -103,12 +103,12 @@ func (engineDetails EngineDetails) WatchChaosContainerForCompletion(experiment *
...
 		}
 
 		expStatus.AwaitedExperimentStatus(experiment.Name, engineDetails.Name, chaosPod.Name)
 		if err := expStatus.PatchChaosEngineStatus(engineDetails, clients); err != nil {
-			return errors.Errorf("Unable to patch ChaosEngine in namespace: %v, error: %v", engineDetails.EngineNamespace, err)
+			return errors.Errorf("unable to patch ChaosEngine in namespace: %v, error: %v", engineDetails.EngineNamespace, err)
```
Comment (rahulchheda) [inline pkg/utils/watchChaosContainer.go:111]: use `fmt.Errorf` to wrap errors.

File: pkg/utils/types.go [hunk 21]
```diff
@@ -123,15 +123,5 @@ func getKubeConfig() (*rest.Config, error) {
...
-		if err != nil {
-			return config, err
-		}
-	}
-	config, err := clientcmd.BuildConfigFromFlags("", *kubeconfig)
-	if err != nil {
-		return config, err
-	}
-	return config, err
+	return clientcmd.BuildConfigFromFlags("", *kubeconfig)
```
Comment (rahulchheda) [inline pkg/utils/types.go:126]: de-referencing pointer, lets try to avoid this, if possible.

**Review by rahulchheda — APPROVED** [review]
Few nit picks.
Approving this, please fix them and merge.

File: tests/runner_test.go (hunk 6 above)
Comment (ispeakc0de) [reply tests/runner_test.go:223]: Yes both are correct

File: pkg/utils/watchChaosContainer.go (hunk 18 above)
Comment (ispeakc0de) [reply pkg/utils/watchChaosContainer.go:73]: we are using errors at every place. Will update in another PR

**Review by rahulchheda — APPROVED** [review]
/lgtm

**Review by ksatchit — APPROVED** [review]

**Pull request merged**